File: sdks/android/Mentat/app/src/main/java/com/mozilla/mentat/TupleResult.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * Copyright 2018 Mozilla
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License. */

package com.mozilla.mentat;

import android.util.Log;
import com.sun.jna.Pointer;

import java.nio.ByteBuffer;
import java.util.Date;
import java.util.UUID;

public class TupleResult extends RustObject {

    public TupleResult(Pointer pointer) {
        this.rawPointer = pointer;
    }

    public TypedValue get(Integer index) {
        this.validate();
        Pointer pointer = JNA.INSTANCE.value_at_index(this.rawPointer, index);
        if (pointer == null) {
            return null;
        }
        return new TypedValue(pointer);
    }

    public Long asLong(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_long(this.rawPointer, index);
    }

    public Long asEntid(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_entid(this.rawPointer, index);
    }

    public String asKeyword(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_kw(this.rawPointer, index);
    }

    public Boolean asBool(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_boolean(this.rawPointer, index) == 0 ? false : true;
    }

    public Double asDouble(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_double(this.rawPointer, index);
    }

    public Date asDate(Integer index) {
        this.validate();
        return new Date(JNA.INSTANCE.value_at_index_as_timestamp(this.rawPointer, index));
    }

    public String asString(Integer index) {
        this.validate();
        return JNA.INSTANCE.value_at_index_as_string(this.rawPointer, index);
    }

    public UUID asUUID(Integer index) {
        this.validate();
        Pointer uuidPtr = JNA.INSTANCE.value_at_index_as_uuid(this.rawPointer, index);
        byte[] bytes = uuidPtr.getByteArray(0, 16);
        ByteBuffer bb = ByteBuffer.wrap(bytes);
        long high = bb.getLong();
        long low = bb.getLong();

        return new UUID(high, low);
    }

    @Override
    public void close() {
        Log.i("TupleResult", "close");
        if (this.rawPointer != null) {
            JNA.INSTANCE.typed_value_list_destroy(this.rawPointer);
        }
    }
}

File: sdks/swift/Mentat/Mentat/Query/TupleResult.swift
/* Copyright 2018 Mozilla
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License. */

import Foundation
import MentatStore

class TupleResult: OptionalRustObject {

    func get(index: Int) -> TypedValue {
        return TypedValue(raw: value_at_index(self.raw!, Int32(index)))
    }

    func asLong(index: Int) -> Int64 {
        return value_at_index_as_long(self.raw!, Int32(index))
    }

    func asEntid(index: Int) -> Int64 {
        return value_at_index_as_entid(self.raw!, Int32(index))
    }

    func asKeyword(index: Int) -> String {
        return String(cString: value_at_index_as_kw(self.raw!, Int32(index)))
    }

    func asBool(index: Int) -> Bool {
        return value_at_index_as_boolean(self.raw!, Int32(index)) == 0 ? false : true
    }

    func asDouble(index: Int) -> Double {
        return value_at_index_as_double(self.raw!, Int32(index))
    }

    func asDate(index: Int) -> Date {
        return Date(timeIntervalSince1970: TimeInterval(value_at_index_as_timestamp(self.raw!, Int32(index))))
    }

    func asString(index: Int) -> String {
        return String(cString: value_at_index_as_string(self.raw!, Int32(index)))
    }

    func asUUID(index: Int) -> UUID? {
        return UUID(uuid: value_at_index_as_uuid(self.raw!, Int32(index)).pointee)
    }

    override func cleanup(pointer: OpaquePointer) {
        typed_value_list_destroy(pointer)
    }
}

class ColResult: TupleResult {
}

class ColResultIterator: OptionalRustObject, IteratorProtocol  {
    typealias Element = TypedValue

    init(iter: OpaquePointer?) {
        super.init(raw: iter)
    }

    func next() -> Element? {
        guard let iter = self.raw,
            let rowPtr = values_iter_next(iter) else {
                return nil
        }
        return TypedValue(raw: rowPtr)
    }

    override func cleanup(pointer: OpaquePointer) {
        typed_value_list_iter_destroy(pointer)
    }
}

extension ColResult: Sequence {
    func makeIterator() -> ColResultIterator {
        defer {
            self.raw = nil
        }
        guard let raw = self.raw else {
            return ColResultIterator(iter: nil)
        }
        let rowIter = values_iter(raw)
        return ColResultIterator(iter: rowIter)
    }
}
